internal/git/gitsync_processor.go

```diff
@@ -67,55 +66,79 @@ func watchRepo(ctx context.Context, repo *v1.RepositoryPath, _ /* namespace */ s
 	}
 
 	// The revision can be a branch, a tag, or a commit hash
-	h, err := r.ResolveRevision(plumbing.Revision(repo.TargetRevision))
+	hash, err := r.ResolveRevision(plumbing.Revision(repo.TargetRevision))
 	if err != nil {
 		logger.Errorw("error resolving the revision", "revision", repo.TargetRevision, "err", err)
 		return err
 	}
 
-	// Retrieving the commit object matching the hash.
-	commit, err := r.CommitObject(*h)
-	if err != nil {
-		logger.Errorw("error checkout the commit", "hash", h.String(), "err", err)
-		return err
-	}
-
-	// Retrieve the tree from the commit.
-	tree, err := commit.Tree()
-	if err != nil {
-		logger.Errorw("error get the commit tree", "err", err)
-		return err
-	}
-
-	if !isRootDir(repo.Path) {
-		// Locate the tree with the given path.
-		tree, err = tree.Tree(repo.Path)
+	val, ok := gitSync.Status.CommitStatus[repo.Name]
+	// Only create the resources for the first time if not created yet.
+	// Otherwise, monitoring with intervals.
+	if !ok {
+		// Retrieving the commit object matching the hash.
+		commit, err := r.CommitObject(*hash)
 		if err != nil {
-			logger.Errorw("error locate the path", "err", err)
+			logger.Errorw("error checkout the commit", "hash", hash.String(), "err", err)
 			return err
 		}
-	}
 
-	// Read all the files under the path and apply each one respectively.
-	err = tree.Files().ForEach(func(f *object.File) error {
-		logger.Debugw("read file", "file_name", f.Name)
-		_, err = f.Contents()
+		// Retrieve the tree from the commit.
+		tree, err := commit.Tree()
 		if err != nil {
-			logger.Errorw("cannot get file content", "filename", f.Name, "err", err)
+			logger.Errorw("error get the commit tree", "err", err)
 			return err
 		}
 
-		// TODO: Apply to the resources
-		return nil
-	})
-	if err != nil {
-		return err
+		if !isRootDir(repo.Path) {
+			// Locate the tree with the given path.
+			tree, err = tree.Tree(repo.Path)
+			if err != nil {
+				logger.Errorw("error locate the path", "err", err)
+				return err
+			}
+		}
+
+		// Read all the files under the path and apply each one respectively.
+		err = tree.Files().ForEach(func(f *object.File) error {
+			logger.Debugw("read file", "file_name", f.Name)
+			_, err = f.Contents()
+			if err != nil {
+				logger.Errorw("cannot get file content", "filename", f.Name, "err", err)
+				return err
+			}
+
+			// TODO: Apply to the resources
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		commitStatus := v1.CommitStatus{
+			Hash:     hash.String(),
+			Synced:   true,
+			SyncTime: metav1.NewTime(time.Now()),
+		}
+		if gitSync.Status.CommitStatus == nil {
+			gitSync.Status.CommitStatus = make(map[string]v1.CommitStatus)
```

Comment:
It's not thread safe to update the central map. So, I guess there are 2 options for dealing with this:
1. Continue to update gitSync.Status.CommitStatus but in a thread safe way
2. Maintain a copy of the gitSync.Status here instead of a reference to the actual object. Each time we want to update GitSync.Status, we call out to do a Get() in order to get a local copy, modify it, and then do an update.

Reply:
Note that for the 2nd option, we don't need to be concerned about the extra API call - since there is a local informer cache that holds the value.

Reply:
I haven't thought about it thoroughly but thinking like we may need to go with the 2nd approach - if we do the 1st approach, we'd need to make sure there's no way that one goroutine's CommitStatus could fail to get updated.

Reply:
Yeah, I think the second is cleaner and the way to go. 👍